### PR TITLE
SONAR-8857 make ws/qualityprofiles/rename organization aware

### PR DESCRIPTION
**Pull request by bartfastiel**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/RenameAction.java [hunk 1]
```diff
@@ -19,23 +19,44 @@
...
+import static org.sonar.server.ws.WsUtils.checkRequest;
+import static org.sonarqube.ws.client.qualityprofile.QualityProfileWsParameters.PARAM_ORGANIZATION;
 
 public class RenameAction implements QProfileWsAction {
 
   private static final String PARAM_PROFILE_NAME = "name";
   private static final String PARAM_PROFILE_KEY = "key";
+  private static final int MAXIMUM_NAME_LENGTH = 100;
 
   private final QProfileFactory profileFactory;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/RenameAction.java]: MAJOR Remove this unused "profileFactory" private field. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1068)

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/RenameAction.java [hunk 2]
```diff
@@ -19,23 +19,44 @@
...
   private final QProfileWsSupport qProfileWsSupport;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/RenameAction.java]: MAJOR Remove this unused "qProfileWsSupport" private field. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1068)

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/RenameAction.java [hunk 3]
```diff
@@ -19,23 +19,44 @@
  */
 package org.sonar.server.qualityprofile.ws;
 
+import com.google.common.annotations.VisibleForTesting;
+import java.util.Optional;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/RenameAction.java]: MINOR Remove this unused import 'java.util.Optional'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 3 minor
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### 3 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [QProfileFactory.java#L183](https://github.com/SonarSource/sonarqube/blob/748202684dc45a9165d6b4ea46605e2c80928ad9/server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileFactory.java#L183): Remove this use of "selectByNameAndLanguage"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
1. ![MINOR][MINOR] [QProfileFactory.java#L197](https://github.com/SonarSource/sonarqube/blob/748202684dc45a9165d6b4ea46605e2c80928ad9/server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileFactory.java#L197): Remove this use of "selectByNameAndLanguages"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
1. ![INFO][INFO] [QProfileFactory.java#L168](https://github.com/SonarSource/sonarqube/blob/748202684dc45a9165d6b4ea46605e2c80928ad9/server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileFactory.java#L168): Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

[INFO]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/RenameAction.java (hunk 3 above)
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/RenameAction.java]: @SonarTech has a valid point

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/RenameAction.java (hunk 1 above)
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/RenameAction.java]: good point

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/RenameAction.java (hunk 2 above)
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/RenameAction.java]: good point

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/RenameAction.java [hunk 4]
```diff
@@ -59,13 +80,36 @@ public void define(WebService.NewController controller) {
...
+    doHandle(newName, profileKey);
+    response.noContent();
+  }
 
-    profileFactory.rename(profileKey, newName);
+  @VisibleForTesting
+  void doHandle(String newName, String profileKey) {
+    checkRequest(StringUtils.isNotBlank(newName), "Name must be set");
+    checkRequest(newName.length() <= MAXIMUM_NAME_LENGTH, String.format("Name is too long (>%d characters)", MAXIMUM_NAME_LENGTH));
+    userSession.checkLoggedIn();
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/RenameAction.java]: you could even do this check right in the `handle` method
no need to process parameters if the user is not logged in

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/RenameAction.java [hunk 5]
```diff
@@ -59,13 +80,36 @@ public void define(WebService.NewController controller) {
...
+        OrganizationDto organization = dbClient.organizationDao().selectByUuid(dbSession, organizationUuid)
+          .orElseThrow(() -> new IllegalStateException("No organization found for uuid "+organizationUuid));
+        String language = qualityProfile.getLanguage();
+        ofNullable(dbClient.qualityProfileDao().selectByNameAndLanguage(organization, newName, language, dbSession))
+          .ifPresent(found -> {throw BadRequestException.create(format("Quality profile already exists: %s", newName));});
+
+        qualityProfile.setName(newName);
+        dbClient.qualityProfileDao().update(dbSession, qualityProfile);
+      }
+      dbSession.commit();
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/RenameAction.java]: no need to commit if no change is done

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/RenameAction.java [hunk 6]
```diff
@@ -59,13 +80,36 @@ public void define(WebService.NewController controller) {
...
 
-    response.noContent();
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      QualityProfileDto qualityProfile = ofNullable(dbClient.qualityProfileDao().selectByKey(dbSession, profileKey))
+        .orElseThrow(() -> new NotFoundException("Quality profile not found: " + profileKey));
+
+      String organizationUuid = qualityProfile.getOrganizationUuid();
+      userSession.checkPermission(ADMINISTER_QUALITY_PROFILES, organizationUuid);
+
+      if (!StringUtils.equals(newName, qualityProfile.getName())) {
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/RenameAction.java]: how about using JDK `Objects.equals`?

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/RenameAction.java [hunk 7]
```diff
@@ -59,13 +80,36 @@ public void define(WebService.NewController controller) {
...
+        OrganizationDto organization = dbClient.organizationDao().selectByUuid(dbSession, organizationUuid)
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/RenameAction.java]: couldn't you use checkFound here?

(even though it's dead code in practice)

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/RenameActionTest.java [hunk 8]
```diff
@@ -50,46 +51,61 @@
...
-      new RenameAction(new QProfileFactory(dbClient), wsSupport)));
+      underTest));
+    organization = db.organizations().insert();
+
+    createProfiles();
   }
 
   @Test
-  public void rename_nominal() throws Exception {
+  public void rename() {
```
Comment (sns-seb) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/RenameActionTest.java]: I don't see a test ensuring that renaming to a name which exists in another organization is actually supported

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/RenameAction.java (hunk 4 above)
Comment (bartfastiel) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/RenameAction.java]: It is easier to test in "doHandle".

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/RenameAction.java (hunk 6 above)
Comment (bartfastiel) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/RenameAction.java]: This code has not been introduced by me, but I fixed it.

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/RenameAction.java (hunk 7 above)
Comment (bartfastiel) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/RenameAction.java]: I'll use it next time.

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/RenameAction.java (hunk 5 above)
Comment (bartfastiel) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/RenameAction.java]: Absolutely. Fixed. :wrench:

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/RenameActionTest.java (hunk 8 above)
Comment (bartfastiel) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/RenameActionTest.java]: I added a test for that. :gift:

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 2 minor
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### 3 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [QProfileFactory.java#L183](https://github.com/SonarSource/sonarqube/blob/f05a87d3c677622b32ebffe11887936fe42f13f7/server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileFactory.java#L183): Remove this use of "selectByNameAndLanguage"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
1. ![MINOR][MINOR] [QProfileFactory.java#L197](https://github.com/SonarSource/sonarqube/blob/f05a87d3c677622b32ebffe11887936fe42f13f7/server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileFactory.java#L197): Remove this use of "selectByNameAndLanguages"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
1. ![INFO][INFO] [QProfileFactory.java#L168](https://github.com/SonarSource/sonarqube/blob/f05a87d3c677622b32ebffe11887936fe42f13f7/server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileFactory.java#L168): Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

[INFO]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

**Pull request merged**